tests/engine/test_validation.py

```diff
@@ -0,0 +1,511 @@
+    class MyComponent(TestComponentWithoutRun):
+        def run(self):
+            pass
+
+    schema = create_test_schema(uses=MyComponent)
+
+    with pytest.raises(GraphSchemaValidationException):
+        validation.validate(schema, language=None, is_train_graph=True)
+
+
+def test_graph_with_fingerprintable_output():
+    class MyComponent(TestComponentWithoutRun):
+        def run(self) -> TrainingData:
+            pass
+
+    schema = create_test_schema(uses=MyComponent)
+
+    validation.validate(schema, language=None, is_train_graph=True)
+
+
+class MyTrainingData(TrainingData):
+    pass
+
+
+def test_graph_with_fingerprintable_output_subclass():
+    class MyComponent(TestComponentWithoutRun):
+        def run(self) -> MyTrainingData:
+            pass
+
+    schema = create_test_schema(uses=MyComponent)
+
+    validation.validate(schema, language=None, is_train_graph=True)
+
+
+def test_graph_constructor_missing():
+    class MyComponent(TestComponentWithoutRun):
+        def run(self) -> TrainingData:
+            pass
+
+    schema = create_test_schema(uses=MyComponent, constructor_name="invalid")
+
+    with pytest.raises(GraphSchemaValidationException):
+        validation.validate(schema, language=None, is_train_graph=True)
+
+
+def test_graph_constructor_config_wrong_type():
+    class MyComponent(TestComponentWithRun):
+        @classmethod
+        def create(
+            cls,
+            config: Dict[int, int],
+            model_storage: ModelStorage,
+            resource: Resource,
+            execution_context: ExecutionContext,
+        ) -> GraphComponent:
+            pass
+
+    schema = create_test_schema(uses=MyComponent)
+
+    with pytest.raises(GraphSchemaValidationException):
+        validation.validate(schema, language=None, is_train_graph=True)
+
+
+def test_graph_constructor_resource_wrong_type():
+    class MyComponent(TestComponentWithRun):
+        @classmethod
+        def create(
+            cls,
+            config: Dict[Text, Any],
+            model_storage: ModelStorage,
+            resource: Dict,
+            execution_context: ExecutionContext,
+        ) -> GraphComponent:
+            pass
+
+    schema = create_test_schema(uses=MyComponent)
+
+    with pytest.raises(GraphSchemaValidationException):
+        validation.validate(schema, language=None, is_train_graph=True)
+
+
+def test_graph_constructor_model_storage_wrong_type():
+    class MyComponent(TestComponentWithRun):
+        @classmethod
+        def create(
+            cls,
+            config: Dict[Text, Any],
+            model_storage: Any,
+            resource: Resource,
+            execution_context: ExecutionContext,
+        ) -> GraphComponent:
+            pass
+
+    schema = create_test_schema(uses=MyComponent)
+
+    with pytest.raises(GraphSchemaValidationException):
+        validation.validate(schema, language=None, is_train_graph=True)
+
+
+def test_graph_constructor_execution_context_wrong_type():
```

Comment:
could parameterize the `test_graph_constructor_..._wrong_type` tests?

Reply:
same as https://github.com/RasaHQ/rasa/pull/9294/files#r685316330